internal/database/postgres.go

```diff
@@ -309,66 +311,125 @@ FROM TempDownsampled;
 }
 
 // TODO: consider returning workstationGroup
-func (conn PostgresConn) LatestData() ([]uplink.GpuStatsUpload, error) {
-	// we pull Uuid twice so we can put one into Stat sample and the other into Info
-	rows, err := conn.db.Query(`SELECT g.Machine, g.Uuid, g.Uuid, g.Name,
-			g.Brand, g.DriverVersion, g.MemoryTotal,
-			s.MemoryUtilisation, s.GpuUtilisation, s.MemoryUsed,
-			s.FanSpeed, s.Temp, s.MemoryTemp, s.GraphicsVoltage,
-			s.PowerDraw, s.GraphicsClock, s.MaxGraphicsClock,
-			s.MemoryClock, s.MaxMemoryClock
+func (conn PostgresConn) LatestData() (broadcast.Workstations, error) {
+	// pull all the machines in, then gpus for those machines
+	// has to all be done in a transaction to avoid race conditions
+	tx, err := conn.db.Begin()
+	if err != nil {
+		return nil, err
+	}
+
+	groups := make(map[string][]broadcast.Workstation)
+	machines, err := tx.Query(`SELECT GroupName, Hostname, CPU, Motherboard,
+		Notes, LastSeen
+		FROM Machines`)
+	if err != nil {
+		return nil, errors.Join(err, tx.Rollback())
+	}
+
+	for machines.Next() {
+		var groupName *string
+		var machine broadcast.Workstation
+		var lastSeen time.Time
+
+		err = machines.Scan(&groupName, &machine.Name, &machine.CPU,
+			&machine.Motherboard, &machine.Notes, &lastSeen)
+		if err != nil {
+			return nil, errors.Join(err, tx.Rollback())
+		}
+
+		// coalesce null and empty group names to default
+		if groupName == nil || strings.TrimSpace(*groupName) == "" {
+			fallback := DefaultGroup
+			groupName = &fallback
+		}
+
+		machine.LastSeen = time.Since(lastSeen)
+		machine.Gpus = nil
+
+		groups[*groupName] = append(groups[*groupName], machine)
+	}
+
+	// check for error whilst iterating, continuing if it's "no results"
+	err = machines.Err()
+	if err != nil && !errors.Is(sql.ErrNoRows, err) {
+		return nil, errors.Join(err, tx.Rollback())
+	}
+
+	// attach gpus to all machines
+	// can't be done in the previous loop because we can't be iterating
+	// through two queries at once
+	for group := range groups {
+		for i, machine := range groups[group] {
+			machine.Gpus, err = getGpus(machine.Name, tx)
+			if err != nil {
+				return nil, errors.Join(err, tx.Rollback())
+			}
+			groups[group][i] = machine
+		}
+	}
+
+	// flatten map
+	var result broadcast.Workstations
+	for groupName, machines := range groups {
+		result = append(result, broadcast.Group{
+			Name:         groupName,
+			Workstations: machines,
+		})
+	}
+
+	return result, tx.Commit()
+}
+
+// get the latest stat for all the gpus on a machine
+func getGpus(host string, tx *sql.Tx) ([]broadcast.GPU, error) {
+	result := make([]broadcast.GPU, 0)
+
+	gpus, err := tx.Query(`SELECT g.Uuid, g.Name, g.Brand,
```

Comment:
Better aliases on this query would be ideal.

Reply:
Just to be clear here, the aliases you're referring to here are how `g` refers to the `Gpus` table and `s` refers to the `Stats` table?